### PR TITLE
Sde Tutorial

### LINKED ISSUE #1987: Add tutorials and examples for SDE
#1884 is about to get merged, we should add a few basic examples at least to 
show how to use it from Python and C++. Also, some standard mechanisms could
be included in a 'stochastic' catalogue.

### TODO

- [x] add a simple example
  - [x] C++
  - [x] Python
- [x] add stochastic catalogue
  - [x] O-U process point source
- [ ] Optional: add tutorial

### PR DESCRIPTION
**Pull request by boeschf**
Implementation of stochastic calcium plasticity curves, originally authored by Sebastian Schmitt @schmitts, and adapted here for Arbor inclusion.
- python example plasticity_stdp.py
- stochastic mechanism calcium_based_synapse.mod
- tutorial

Closes #1987

### CODE REVIEW

File: doc/tutorial/calcium_stdp_curve.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
+.. _tutorial_calcium_stpd_curve:
+
+Spike Timing-dependent Plasticity Curve
+=======================================
+
+This tutorial uses a single cell and reproduces `this Brian2 example
+<https://brian2.readthedocs.io/en/latest/examples/frompapers.Graupner_Brunel_2012.html>`_.  We aim
+to reproduce a spike timing-dependent plastivity curve which arises form stochastic calcium-based
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:8]: typo: form

File: doc/tutorial/calcium_stdp_curve.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                     \rho (1 - \rho) (\rho^\ast - \rho)
+                     -\gamma_p (1-\rho) H\left(c - \theta_p \right)
+                     + \gamma_d \rho H\left(c - \theta_d \right) \right ]
+                     + N, \\
+   N &= \frac{\sigma}{\sqrt{\tau}} \sqrt{H\left( c - \theta_p \right)
+        + H\left( c - \theta_d \right)} W.
+   \end{align*}
+
+Here, the sums over :math:`i` and :math:`j` represent the contributions from all pre- and
+postsynaptic spikes, respectively, with :math:`C_{pre}` and :math:`C_{pre}` denoting the jumps in
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:34]: typo: pre and post

File: doc/tutorial/calcium_stdp_curve.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+concentration after a spike. The jump after the presynaptic spike is delayed by :math:`D`.  The
+Calcium decay time is assumed to be much faster than the synaptic time scale,
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:36]: typo: make lower case calcium

File: doc/tutorial/calcium_stdp_curve.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+The main difference compared to a deterministic (ODE) description is the additional `WHITE_NOISE`
+block,
+
+.. code:: none
+
+   WHITE_NOISE {
+       W
+   }
+
+which declares the white noise process :math:`W`, and the specification pf the `stochastic` solver
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:65]: typo: pf -> of

File: doc/tutorial/calcium_stdp_curve.rst [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+method,
+
+.. code:: none
+
+   BREAKPOINT {
+       SOLVE state METHOD stochastic
+   }
+
+This is sufficient to inform Arbor about the stochasticity of the mechanism.
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:74]: can we add a link to an explanation of what kind of solver `stochastic` implements? https://docs.arbor-sim.org/en/latest/dev/numerics.html

File: doc/tutorial/calcium_stdp_curve.rst [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+Next we set the simulation parameters in order to reproduce the plasticity curve:
+
+.. literalinclude:: ../../python/example/calcium_stdp.py
+   :language: python
+   :lines: 20-41
+
+The time lag resolution, together with the maximum time lag, determine the number of cases we want
+to simulate. For each such case, however, we need to run many simulations in order to get a
+statistically meaningful result. The number of simulations per case is given by the ensemble size
+and the initial conditions. In our case, we have to inital states, :math:`\rho(0)=0` and
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:99]: typo: to -> two

File: doc/tutorial/calcium_stdp_curve.rst [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+:math:`\rho(0)=1`, and for each initial state we want to run :math:`100` simulations. This can
+easiest be achieved by placing :math:`100` synapses per initial state onto the cell.
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:101]: maybe add ... and allows to parallize?

File: doc/tutorial/calcium_stdp_curve.rst [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+Next we set up the simulation logic:
+
+.. literalinclude:: ../../python/example/calcium_stdp.py
+   :language: python
+   :lines: 106-134
+
+The pre- and postsynaptic events are generated at explicit schedules, where the presynaptic event
+is shifted in time by :math:`D -\text{time lag}` with respect to the presynaptic event, which in
+turn is generated regularly with the frequency :math:`f`. The postsynaptic events are driven by the
+deterministic synapse with weight `1.0`, while the presynaptic events are generated at the
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:137]: maybe add: arbitrary, but large, weight ...?

File: doc/tutorial/calcium_stdp_curve.rst [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+Now, we need to initialize the simulation, register a probe and run the simulation:
+
+.. literalinclude:: ../../python/example/calcium_stdp.py
+   :language: python
+   :lines: 139-154
+
+Since we are interested in the long-term average value, we only query the probe at the end of the
+simulation.
+
+After the simulation is finished, we caclculate the change in synaptic strength by evaluating the
```
Comment (schmitts) [inline doc/tutorial/calcium_stdp_curve.rst:155]: typo in calculate

**Review by schmitts — CHANGES_REQUESTED** [review]
Thanks for the writeup! Minor fixes and suggested additions.

File: doc/tutorial/calcium_stdp_curve.rst (hunk 5 above)
Comment (boeschf) [reply doc/tutorial/calcium_stdp_curve.rst]: Yes you are right, that should be mentioned. I added links to the longer explanations of how we treat SDEs but also added an entry to the `dev/numerics.rst` - this document was actually commented out, so I re-added it to the TOC.

File: doc/tutorial/calcium_stdp_curve.rst (hunk 7 above)
Comment (boeschf) [reply doc/tutorial/calcium_stdp_curve.rst]: true, added some more explanation

File: doc/tutorial/calcium_stdp_curve.rst (hunk 8 above)
Comment (boeschf) [reply doc/tutorial/calcium_stdp_curve.rst:143]: ok, expanded explanation

**Review by schmitts — APPROVED** [review]
LGTM

**Pull request merged**